File: XIVRaidToolsPlugin/Icons.cs
using System.Numerics;
using Dalamud.Bindings.ImGui;

namespace XIVRaidToolsPlugin;

// Vector re-draws of kefka-says/icons.js's inline SVGs onto an ImGui draw
// list, at the same 44x44 viewBox the source SVGs use. No texture/PNG
// pipeline needed — everything here is lines, circles, and a couple of
// manually-triangulated fills, scaled to whatever box the caller draws into.
public delegate void Icon(ImDrawListPtr dl, Vector2 topLeft, float size, uint color);

public static class Icons
{
    private static Vector2 P(Vector2 origin, float s, float x, float y) => origin + new Vector2(x, y) * s;

    // Shaft + a 2-segment arrowhead chevron meeting at one point is a 3-way
    // junction (branch), which can't be a single continuous path — a path is
    // a linear chain, it can't fork. The chevron's own 2 segments CAN be one
    // continuous path (no seam between them), but the shaft joining that
    // same point is necessarily a second, separate stroke — so a small
    // filled circle "cap" sits over the junction, sized to the stroke width,
    // masking whatever seam the two independently anti-aliased strokes would
    // otherwise leave right at the corner.
    private static void ArmWithChevron(ImDrawListPtr dl, Vector2 o, float s, uint color,
        float x1, float y1, float x2, float y2, float ax, float ay, float bx, float by, float thickness)
    {
        dl.AddLine(P(o, s, x1, y1), P(o, s, x2, y2), color, thickness);
        dl.PathLineTo(P(o, s, ax, ay));
        dl.PathLineTo(P(o, s, x2, y2));
        dl.PathLineTo(P(o, s, bx, by));
        dl.PathStroke(color, ImDrawFlags.None, thickness);
        dl.AddCircleFilled(P(o, s, x2, y2), thickness / 2f, color);
    }

    public static void Spread(ImDrawListPtr dl, Vector2 o, float size, uint color)
    {
        var s = size / 44f;
        var t = 2.5f * s;
        ArmWithChevron(dl, o, s, color, 22, 22, 36, 8, 29, 8, 36, 15, t);
        ArmWithChevron(dl, o, s, color, 22, 22, 8, 36, 15, 36, 8, 29, t);
        ArmWithChevron(dl, o, s, color, 22, 22, 8, 8, 15, 8, 8, 15, t);
        ArmWithChevron(dl, o, s, color, 22, 22, 36, 36, 29, 36, 36, 29, t);
    }

    public static void Stack(ImDrawListPtr dl, Vector2 o, float size, uint color)
    {
        var s = size / 44f;
        var t = 2.5f * s;
        // Same junction as Spread's arms, just pointing inward (chevron end
        // is the shaft's START here, x1/y1, rather than its end).
        ArmWithChevron(dl, o, s, color, 25, 19, 36, 8, 29, 8, 36, 15, t);
        ArmWithChevron(dl, o, s, color, 19, 25, 8, 36, 15, 36, 8, 29, t);
        ArmWithChevron(dl, o, s, color, 19, 19, 8, 8, 15, 8, 8, 15, t);
        ArmWithChevron(dl, o, s, color, 25, 25, 36, 36, 29, 36, 36, 29, t);
        dl.AddCircleFilled(P(o, s, 22, 22), 4 * s, color);
    }

    public static void Still(ImDrawListPtr dl, Vector2 o, float size, uint color)
    {
        var s = size / 44f;
        dl.AddRectFilled(P(o, s, 10, 9), P(o, s, 20, 35), color, 2 * s);
        dl.AddRectFilled(P(o, s, 24, 9), P(o, s, 34, 35), color, 2 * s);
    }

    public static void Move(ImDrawListPtr dl, Vector2 o, float size, uint color)
    {
        var s = size / 44f;
        dl.AddTriangleFilled(P(o, s, 9, 5), P(o, s, 9, 39), P(o, s, 39, 22), color);
    }

    // Eye lens approximated as two mirrored bezier arcs (no AddEllipse in
    // this binding) plus a filled pupil, same silhouette as icons.js's path.
    private static void EyeOutline(ImDrawListPtr dl, Vector2 o, float s, uint color, float thickness)
    {
        var left = P(o, s, 4, 22);
        var right = P(o, s, 40, 22);
        var topCtrl = P(o, s, 22, 6);
        var botCtrl = P(o, s, 22, 38);
        dl.PathLineTo(left);
        dl.PathBezierQuadraticCurveTo(topCtrl, right);
        dl.PathBezierQuadraticCurveTo(botCtrl, left);
        dl.PathStroke(color, ImDrawFlags.Closed, thickness);
    }

    public static void GazeIn(ImDrawListPtr dl, Vector2 o, float size, uint color)
    {
        var s = size / 44f;
        EyeOutline(dl, o, s, color, 2.5f * s);
        dl.AddCircleFilled(P(o, s, 22, 22), 7 * s, color);
    }

    public static void GazeOut(ImDrawListPtr dl, Vector2 o, float size, uint color)
    {
        var s = size / 44f;
        EyeOutline(dl, o, s, ScaleAlpha(color, 0.3f), 2.5f * s);
        dl.AddCircleFilled(P(o, s, 22, 22), 7 * s, ScaleAlpha(color, 0.3f));
        dl.AddLine(P(o, s, 8, 8), P(o, s, 36, 36), Theme.U32(Theme.Fake), 3f * s);
    }

    public static void Circle(ImDrawListPtr dl, Vector2 o, float size, uint color)
    {
        var s = size / 44f;
        dl.AddCircleFilled(P(o, s, 22, 22), 19 * s, ScaleAlpha(color, 0.85f));
    }

    public static void Donut(ImDrawListPtr dl, Vector2 o, float size, uint color)
    {
        var s = size / 44f;
        dl.AddCircle(P(o, s, 22, 22), 14.5f * s, ScaleAlpha(color, 0.85f), 0, 9 * s);
    }

    // Stroked zigzag rather than a filled hexagon: the previous version
    // triangulated the classic bolt silhouette into 3 manual triangles, but
    // adjoining triangle edges each anti-alias independently, leaving a
    // visible seam through the shape — much more noticeable at the small
    // sizes these icons render at (button icons, ~15-20px) than it was in
    // the original at-scale mockup. This zigzag is a genuinely sequential
    // path (unlike Spread/Stack's 3-way arm junctions), so it really can be
    // ONE continuous PathStroke — the previous version still used 3 separate
    // AddLine calls despite the comment here claiming otherwise, and had the
    // exact same joint-seam bug as the fill version it replaced.
    private static void BoltPath(ImDrawListPtr dl, Vector2 o, float s, uint color, float thickness)
    {
        dl.PathLineTo(P(o, s, 27, 3));
        dl.PathLineTo(P(o, s, 14, 23));
        dl.PathLineTo(P(o, s, 25, 23));
        dl.PathLineTo(P(o, s, 15, 41));
        dl.PathStroke(color, ImDrawFlags.None, thickness);
    }

    public static void Thunder(ImDrawListPtr dl, Vector2 o, float size, uint color)
    {
        var s = size / 44f;
        BoltPath(dl, o, s, color, 4f * s);
    }

    public static void ThunderFake(ImDrawListPtr dl, Vector2 o, float size, uint color)
    {
        var s = size / 44f;
        BoltPath(dl, o, s, ScaleAlpha(color, 0.3f), 4f * s);
        dl.AddLine(P(o, s, 6, 6), P(o, s, 38, 38), Theme.U32(Theme.Fake), 3.5f * s);
    }

    // Inferno (flame) and Tsunami (wave) — for the compact Floor AOE #1 Type
    // buttons, replacing the "Inferno"/"Tsunami" text buttons. Flame reuses
    // Droplet's triangle+circle construction (simple, no seam risk) with an
    // asymmetric lean so it doesn't read as a second water drop; Wave is a
    // stroked double-arc (no fill, same anti-seam reasoning as the bolt).
    public static void Flame(ImDrawListPtr dl, Vector2 o, float size, uint color)
    {
        var s = size / 44f;
        dl.AddTriangleFilled(P(o, s, 27, 3), P(o, s, 8, 29), P(o, s, 35, 24), color);
        dl.AddCircleFilled(P(o, s, 21, 29), 13 * s, color);
    }

    public static void Wave(ImDrawListPtr dl, Vector2 o, float size, uint color)
    {
        var s = size / 44f;
        var t = 3f * s;
        dl.PathLineTo(P(o, s, 4, 28));
        dl.PathBezierQuadraticCurveTo(P(o, s, 14, 14), P(o, s, 22, 28));
        dl.PathStroke(color, ImDrawFlags.None, t);
        dl.PathLineTo(P(o, s, 22, 28));
        dl.PathBezierQuadraticCurveTo(P(o, s, 30, 42), P(o, s, 40, 28));
        dl.PathStroke(color, ImDrawFlags.None, t);
    }

    private static void SnowflakePath(ImDrawListPtr dl, Vector2 o, float s, uint color, float thickness)
    {
        void L(float x1, float y1, float x2, float y2) => dl.AddLine(P(o, s, x1, y1), P(o, s, x2, y2), color, thickness);
        L(22, 4, 22, 40);
        L(4, 22, 40, 22);
        L(9, 9, 35, 35);
        L(35, 9, 9, 35);
        L(22, 4, 16, 11); L(22, 4, 28, 11);
        L(22, 40, 16, 33); L(22, 40, 28, 33);
        L(4, 22, 11, 16); L(4, 22, 11, 28);
        L(40, 22, 33, 16); L(40, 22, 33, 28);
    }

    public static void Blizzard(ImDrawListPtr dl, Vector2 o, float size, uint color)
    {
        SnowflakePath(dl, o, size / 44f, color, 2.5f * size / 44f);
    }

    public static void BlizzardFake(ImDrawListPtr dl, Vector2 o, float size, uint color)
    {
        var s = size / 44f;
        SnowflakePath(dl, o, s, ScaleAlpha(color, 0.3f), 2.5f * s);
        dl.AddLine(P(o, s, 6, 6), P(o, s, 38, 38), Theme.U32(Theme.Fake), 3.5f * s);
    }

    private static uint ScaleAlpha(uint packedColor, float alpha)
    {
        var a = (byte)(((packedColor >> 24) & 0xFF) * alpha);
        return (packedColor & 0x00FFFFFF) | ((uint)a << 24);
    }

    // Not from icons.js — added for the compact icon-only Real/Fake and
    // debuff-selection buttons (no equivalent in the webapp, which has room
    // for text labels; the in-game window doesn't).
    public static void Check(ImDrawListPtr dl, Vector2 o, float size, uint color)
    {
        // One continuous path, not two separate AddLine calls meeting at
        // (18,34) — two independent line draws can leave a visible seam/gap
        // at their shared corner since each is anti-aliased on its own; a
        // single stroked path has no such joint.
        var s = size / 44f;
        var t = 3.5f * s;
        dl.PathLineTo(P(o, s, 8, 24));
        dl.PathLineTo(P(o, s, 18, 34));
        dl.PathLineTo(P(o, s, 37, 10));
        dl.PathStroke(color, ImDrawFlags.None, t);
    }

    public static void Cross(ImDrawListPtr dl, Vector2 o, float size, uint color)
    {
        var s = size / 44f;
        var t = 3.5f * s;
        dl.AddLine(P(o, s, 10, 10), P(o, s, 34, 34), color, t);
        dl.AddLine(P(o, s, 34, 10), P(o, s, 10, 34), color, t);
    }

    public static void Droplet(ImDrawListPtr dl, Vector2 o, float size, uint color)
    {
        var s = size / 44f;
        dl.AddTriangleFilled(P(o, s, 22, 4), P(o, s, 9, 27), P(o, s, 35, 27), color);
        dl.AddCircleFilled(P(o, s, 22, 27), 13 * s, color);
    }

    public static void Bomb(ImDrawListPtr dl, Vector2 o, float size, uint color)
    {
        var s = size / 44f;
        dl.AddLine(P(o, s, 28, 14), P(o, s, 34, 6), color, 3f * s);
        dl.AddCircleFilled(P(o, s, 35, 5), 3f * s, color);
        dl.AddCircleFilled(P(o, s, 20, 27), 14 * s, color);
    }
}

File: kefka-says/icons.js
const IC = {
  spread: `<svg viewBox="0 0 44 44" fill="none" stroke="currentColor" stroke-width="2.5" stroke-linecap="round" stroke-linejoin="round">
    <line x1="22" y1="22" x2="36" y2="8"/><polyline points="29,8 36,8 36,15"/>
    <line x1="22" y1="22" x2="8" y2="36"/><polyline points="15,36 8,36 8,29"/>
    <line x1="22" y1="22" x2="8" y2="8"/><polyline points="15,8 8,8 8,15"/>
    <line x1="22" y1="22" x2="36" y2="36"/><polyline points="29,36 36,36 36,29"/>
  </svg>`,

  stack: `<svg viewBox="0 0 44 44" fill="none" stroke="currentColor" stroke-width="2.5" stroke-linecap="round" stroke-linejoin="round">
    <line x1="36" y1="8" x2="25" y2="19"/><polyline points="36,15 36,8 29,8"/>
    <line x1="8" y1="36" x2="19" y2="25"/><polyline points="8,29 8,36 15,36"/>
    <line x1="8" y1="8" x2="19" y2="19"/><polyline points="8,15 8,8 15,8"/>
    <line x1="36" y1="36" x2="25" y2="25"/><polyline points="36,29 36,36 29,36"/>
    <circle cx="22" cy="22" r="4" fill="currentColor" stroke="none"/>
  </svg>`,

  still: `<svg viewBox="0 0 44 44">
    <rect x="10" y="9" width="10" height="26" rx="2" fill="currentColor"/>
    <rect x="24" y="9" width="10" height="26" rx="2" fill="currentColor"/>
  </svg>`,

  move: `<svg viewBox="0 0 44 44">
    <polygon points="9,5 9,39 39,22" fill="currentColor"/>
  </svg>`,

  gazeIn: `<svg viewBox="0 0 44 44" fill="none">
    <path d="M4,22 Q22,6 40,22 Q22,38 4,22 Z" stroke="currentColor" stroke-width="2.5"/>
    <circle cx="22" cy="22" r="7" fill="currentColor"/>
    <circle cx="25" cy="19" r="2.5" fill="white" opacity="0.4"/>
  </svg>`,

  gazeOut: `<svg viewBox="0 0 44 44" fill="none">
    <g opacity="0.3">
      <path d="M4,22 Q22,6 40,22 Q22,38 4,22 Z" stroke="currentColor" stroke-width="2.5"/>
      <circle cx="22" cy="22" r="7" fill="currentColor"/>
    </g>
    <line x1="8" y1="8" x2="36" y2="36" stroke="#cc5555" stroke-width="3" stroke-linecap="round"/>
  </svg>`,

  circle: `<svg viewBox="0 0 44 44">
    <circle cx="22" cy="22" r="19" fill="currentColor" opacity="0.85"/>
  </svg>`,

  donut: `<svg viewBox="0 0 44 44">
    <circle cx="22" cy="22" r="14.5" fill="none" stroke="currentColor" stroke-width="9" opacity="0.85"/>
  </svg>`,

  thunder: `<svg viewBox="0 0 44 44" fill="currentColor">
    <polygon points="27,2 13,25 23,25 17,42 31,19 21,19"/>
  </svg>`,

  thunderFake: `<svg viewBox="0 0 44 44">
    <polygon points="27,2 13,25 23,25 17,42 31,19 21,19" fill="currentColor" opacity="0.3"/>
    <line x1="6" y1="6" x2="38" y2="38" stroke="#cc5555" stroke-width="3.5" stroke-linecap="round"/>
  </svg>`,

  blizzard: `<svg viewBox="0 0 44 44" stroke="currentColor" stroke-width="2.5" stroke-linecap="round" fill="none">
    <line x1="22" y1="4" x2="22" y2="40"/>
    <line x1="4" y1="22" x2="40" y2="22"/>
    <line x1="9" y1="9" x2="35" y2="35"/>
    <line x1="35" y1="9" x2="9" y2="35"/>
    <line x1="22" y1="4" x2="16" y2="11"/><line x1="22" y1="4" x2="28" y2="11"/>
    <line x1="22" y1="40" x2="16" y2="33"/><line x1="22" y1="40" x2="28" y2="33"/>
    <line x1="4" y1="22" x2="11" y2="16"/><line x1="4" y1="22" x2="11" y2="28"/>
    <line x1="40" y1="22" x2="33" y2="16"/><line x1="40" y1="22" x2="33" y2="28"/>
  </svg>`,

  blizzardFake: `<svg viewBox="0 0 44 44" fill="none">
    <g stroke="currentColor" stroke-width="2.5" stroke-linecap="round" opacity="0.3">
      <line x1="22" y1="4" x2="22" y2="40"/>
      <line x1="4" y1="22" x2="40" y2="22"/>
      <line x1="9" y1="9" x2="35" y2="35"/>
      <line x1="35" y1="9" x2="9" y2="35"/>
      <line x1="22" y1="4" x2="16" y2="11"/><line x1="22" y1="4" x2="28" y2="11"/>
      <line x1="22" y1="40" x2="16" y2="33"/><line x1="22" y1="40" x2="28" y2="33"/>
      <line x1="4" y1="22" x2="11" y2="16"/><line x1="4" y1="22" x2="11" y2="28"/>
      <line x1="40" y1="22" x2="33" y2="16"/><line x1="40" y1="22" x2="33" y2="28"/>
    </g>
    <line x1="6" y1="6" x2="38" y2="38" stroke="#cc5555" stroke-width="3.5" stroke-linecap="round"/>
  </svg>`,

  // Icon-mode button glyphs (not derived-status icons like the above — these
  // replace the plain "Real"/"Fake"/"Water"/etc. button text). Ported from
  // the Dalamud plugin's Icons.cs; unlike ImGui's draw list, SVG doesn't
  // double-blend overlapping opaque shapes, so the plugin's "seam" concerns
  // don't apply here — the multi-shape constructions (droplet, flame) are
  // safe as-is.
  check: `<svg viewBox="0 0 44 44" fill="none" stroke="currentColor" stroke-width="3.5" stroke-linecap="round" stroke-linejoin="round">
    <polyline points="8,24 18,34 37,10"/>
  </svg>`,

  cross: `<svg viewBox="0 0 44 44" fill="none" stroke="currentColor" stroke-width="3.5" stroke-linecap="round">
    <line x1="10" y1="10" x2="34" y2="34"/>
    <line x1="34" y1="10" x2="10" y2="34"/>
  </svg>`,

  droplet: `<svg viewBox="0 0 44 44" fill="currentColor">
    <path d="M22,4 L9,27 A13,13 0 1 0 35,27 Z"/>
  </svg>`,

  bomb: `<svg viewBox="0 0 44 44">
    <line x1="28" y1="14" x2="34" y2="6" stroke="currentColor" stroke-width="3" stroke-linecap="round"/>
    <circle cx="35" cy="5" r="3" fill="currentColor"/>
    <circle cx="20" cy="27" r="14" fill="currentColor"/>
  </svg>`,

  flame: `<svg viewBox="0 0 44 44" fill="currentColor">
    <polygon points="27,3 8,29 34,24"/>
    <circle cx="21" cy="29" r="13"/>
  </svg>`,

  wave: `<svg viewBox="0 0 44 44" fill="none" stroke="currentColor" stroke-width="3" stroke-linecap="round">
    <path d="M4,28 Q14,14 22,28 Q30,42 40,28"/>
  </svg>`,
};
